### PR TITLE
MINIFICPP-1116 - Fix error reporting in C2Agent

### PR DESCRIPTION
**Pull request by szaszm**
- Missing CoapProtocol will now result in a fatal error (exception)
instead of segfault. Error reporting is improved.
- C2Agent::protocol_ is now properly deleted
- remove dead code
- avoid unnecessary string copies in loop

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by szaszm** [comment]
The AppVeyor failure does not seem to be related to the changes in this PR. Is it unrelated or am I missing something?

File: libminifi/include/c2/C2Agent.h [hunk 1]
```diff
@@ -57,7 +57,8 @@ class C2Agent : public state::UpdateController, public state::response::Response
 
   C2Agent(const std::shared_ptr<core::controller::ControllerServiceProvider> &controller, const std::shared_ptr<state::StateMonitor> &updateSink, const std::shared_ptr<Configure> &configure);
 
-  virtual ~C2Agent() {
+  ~C2Agent() noexcept override {
```
Comment (bakaid) [inline libminifi/include/c2/C2Agent.h]: In libminifi code we currently don't use override, as there were some problems with it on older systems and it has not yet been thoroughly verified that it works OK now. (In extensions, it is okay.)

File: libminifi/src/c2/C2Agent.cpp [hunk 2]
```diff
@@ -154,25 +155,24 @@ void C2Agent::configure(const std::shared_ptr<Configure> &configure, bool reconf
...
+      logger_->log_warn("Class %s not found", clazz);
       protocol = core::ClassLoader::getDefaultClassLoader().instantiateRaw("CoapProtocol", "CoapProtocol");
       if (!protocol) {
-        logger_->log_info("Attempted to load CoapProtocol. To enable C2, please specify an active protocol for this agent.");
-        return;
-      } else {
-        logger_->log_info("Class is CoapProtocol");
+        const char* errmsg = "Attempted to load CoapProtocol. To enable C2, please specify an active protocol for this agent.";
+        logger_->log_error(errmsg);
+        throw std::domain_error{ errmsg };
```
Comment (bakaid) [inline libminifi/src/c2/C2Agent.cpp]: I think it would be better to use a minifi::Exception here.

File: libminifi/src/c2/C2Agent.cpp [hunk 3]
```diff
@@ -232,7 +232,7 @@ void C2Agent::configure(const std::shared_ptr<Configure> &configure, bool reconf
   if (configure->get("nifi.c2.agent.heartbeat.reporter.classes", "c2.agent.heartbeat.reporter.classes", heartbeat_reporters)) {
     std::vector<std::string> reporters = utils::StringUtils::split(heartbeat_reporters, ",");
     std::lock_guard<std::mutex> lock(heartbeat_mutex);
-    for (auto reporter : reporters) {
+    for (const auto& reporter : reporters) {
```
Comment (bakaid) [inline libminifi/src/c2/C2Agent.cpp]: 👍

File: libminifi/src/c2/C2Agent.cpp [hunk 4]
```diff
@@ -154,25 +155,24 @@ void C2Agent::configure(const std::shared_ptr<Configure> &configure, bool reconf
...
+        throw std::domain_error{ errmsg };
       }
+
+      logger_->log_info("Class is CoapProtocol");
     }
-    C2Protocol *old_protocol = protocol_.exchange(dynamic_cast<C2Protocol*>(protocol));
 
-    protocol_.load()->initialize(controller_, configuration_);
+    // Since !reconfigure, the call comes from the ctor and protocol_ is null, therefore no delete is necessary
+    protocol_.exchange(dynamic_cast<C2Protocol *>(protocol));
```
Comment (bakaid) [inline libminifi/src/c2/C2Agent.cpp]: Nice catch!

File: libminifi/include/c2/C2Agent.h (hunk 1 above)
Comment (szaszm) [reply libminifi/include/c2/C2Agent.h]: fixed in e108d27

File: libminifi/src/c2/C2Agent.cpp (hunk 2 above)
Comment (szaszm) [reply libminifi/src/c2/C2Agent.cpp]: fixed in e108d27

**Pull request closed without merging**